entity-framework/core/testing/choosing-a-testing-strategy.md

```diff
@@ -0,0 +1,112 @@
+
+The following diagram compares the database fake pattern (SQLite/InMemory) with the repository pattern:
+
+![Comparison of fake provider with repository pattern](_static/fake-provider-and-repository-pattern.png)
+
+Since LINQ queries are no longer part of testing, you can directly provide query results to your application. Put another way, the previous approaches roughly allow stubbing out *query inputs* (e.g. replacing SQL Server tables with an in-memory one), but then still execute the actual query operators in-memory; the repository pattern, in contrast, allows you to stub out *query outputs* directly, allowing for far more powerful and focused unit testing.
+
+One possible disadvantage of the repository pattern is that it requires you to architect your application accordingly: application code can no longer execute LINQ queries directly, but must instead call into a repository method, which executes the query (and is stubbed in testing). This can make for a somewhat heavier application architecture, but can also have other advantages: all data access code is concentrated in one place rather than being spread around the application, and if your application needs to support more than one database, the repository abstraction can be very helpful for tweaking queries across providers.
+
+For an example of a unit testing with a repository, see the [unit testing page](xref:core/testing/unit-testing#repository-pattern).
+
+## Overall comparison
+
+The following table provides a quick, comparative view of the different testing techniques, and shows which functionality can be tested under which approach:
+
+Feature                             | InMemory     | SQLite in-memory          | Mock DbContext | Repository pattern | Integration testing
+----------------------------------- | ------------ | ------------------------- | -------------- | ------------------ | -------------------
+Test double type                    | Fake         | Fake                      | Mock/stub      | Mock/stub          | Real, no double
+Raw SQL?                            | No           | Depends                   | No             | Yes                | Yes
+Transactions?                       | No (ignored) | Yes                       | Yes            | Yes                | Yes
+Provider-specific translations?     | No           | No                        | No             | Yes                | Yes
+Exact query behavior?               | Depends      | Depends                   | Depends        | Yes                | Yes
+Requires special code architecture? | No           | No                        | No             | Yes                | No
```

Comment:
```suggestion
Requires special code architecture? | No           | No                        | No             | Yes                | No
Restricts the use of LINQ | No | No | No | Yes | No
```

Reply:
That's an extreme formulation :) I think we're trying to say that all LINQ must be encapsulated behind Enumerable-returning methods (which is what I tried to convey with "Requires special code architecture").

I've replaced "Requires special code architecture" (which was supposed to convey that) with "Can use LINQ anywhere in the application?", where the repository pattern is the only one where the answer is "no" (also added a footnote). Let me know if that seems sufficient.